src/funcx_common/redis_task.py

```diff
@@ -0,0 +1,184 @@
+import typing as t
+
+from .redis import (
+    INT_SERDE,
+    JSON_SERDE,
+    FuncxRedisEnumSerde,
+    HasRedisFieldsMeta,
+    RedisField,
+)
+from .tasks import InternalTaskState, TaskProtocol, TaskState
+
+try:
+    import redis
+
+    has_redis = True
+except ImportError:
+    has_redis = False
+
+
+class RedisTask(TaskProtocol, metaclass=HasRedisFieldsMeta):
+    """
+    ORM-esque class to wrap access to properties of tasks.
+
+    Creation:
+      Create a new task by instantiating this class, e.g.
+      >>> RedisTask(reids_client, "foo_id")
+
+    Loading:
+      Read a task from storage using the `load()` classmethod, e.g.
+      >>> RedisTask.load(redis_client, "foo_id")
+
+    Creation and Loading will each check `RedisTask.exists` when invoked. If the task_id
+    already exists on creation, or does not exist on load, a ValueError will be raised.
+    For example:
+      >>> RedisTask.load(redis_client, "foo_id")
+      Traceback (most recent call last):
+        ...
+      ValueError: Cannot load task foo_id: does not exist
+      >>> RedisTask(reids_client, "foo_id")  # ok
+      >>> RedisTask(reids_client, "foo_id")
+      Traceback (most recent call last):
+        ...
+      ValueError: Conflict. Cannot create task foo_id: already exists
+
+    This class provides various task fields as descriptors via RedisField, and is
+    responsible for de/serializing various data from/to hstore in Redis.
+
+    There are several elements of this pattern of use which need to be fixed. It is
+    important to be aware of the following:
+    - there is currently no use of Redis transactions, so nothing is atomic
+    - Each time a field descriptor is accessed, it is read, returned, and discarded.
+      Reading a field multiple times, even in a single python statement, is vulnerable
+      to data races
+    - no field requirements or validity are enforced -- reading a field can raise an
+      error if bad data were written to Redis
+    - each field is read individually, which can be inefficient and inconsistent (vs
+      getall or setall semantics)
+    """
+
+    # 2 weeks in seconds
+    DEFAULT_TTL: t.ClassVar[int] = 1209600
+
+    status = t.cast(TaskState, RedisField(serde=FuncxRedisEnumSerde(TaskState)))
+    internal_status = t.cast(
+        InternalTaskState, RedisField(serde=FuncxRedisEnumSerde(InternalTaskState))
+    )
+    user_id = t.cast(int, RedisField(serde=INT_SERDE))
+    function_id = t.cast(t.Optional[str], RedisField())
+    endpoint = t.cast(t.Optional[str], RedisField())
+    container = t.cast(str, RedisField())
```

Comment:
Should container be `t.cast(t.Optional[str], RedisField())`?

Reply:
Yes, that seems like a good correction.

The behavior of `RedisField` makes everything (in fact) `Optional`, but we can say that if some fields (e.g. `internal_status`) are not set that the task is "malformed". So any type used in these casts which is _not_ `Optional[...]` is something of a white lie.
I'd like to give `RedisField` a behavior `required=True` which can make those types more accurate.

Reply:
I'm actually going in the opposite direction, having checked the web service. `container` is always specified when a new task is created.
I'm therefore going to not mark it optional and add a comment that it is one of the required fields.

This and the above note on `user_id` are both reasonable in suggesting the use of `Optional`, as that's what's really returned from `RedisField`. But this is a case where I think pragmatism wins out: annotating these as non-optional both correctly encodes the _idea_ that they will never be null and avoids unnecessary checks in order to pacify mypy in consuming code (e..g `if task.user_id is None: raise ...`).